posthog/models/web_preaggregated/sql.py

```diff
@@ -0,0 +1,295 @@
+    PARTITION BY toYYYYMM(day_bucket)
+    ORDER BY {order_by}
+    """
+
+
+def DISTRIBUTED_TABLE_TEMPLATE(dist_table_name, base_table_name, columns):
+    return f"""
+    CREATE OR REPLACE TABLE {dist_table_name} ON CLUSTER '{CLICKHOUSE_CLUSTER}'
+    (
+        day_bucket DateTime,
+        team_id UInt64,
+        host String,
+        device_type String,
+        {columns}
+    ) ENGINE = Distributed('{CLICKHOUSE_CLUSTER}', '{CLICKHOUSE_DATABASE}', {base_table_name}, rand())
+    """
+
+
+# Column definitions
+WEB_OVERVIEW_METRICS_COLUMNS = """
+    persons_uniq_state AggregateFunction(uniq, UUID),
+    sessions_uniq_state AggregateFunction(uniq, String),
+    pageviews_count_state AggregateFunction(sum, UInt64),
+    total_session_duration_state AggregateFunction(sum, Int64),
+    total_bounces_state AggregateFunction(sum, UInt64)
+"""
+
+WEB_STATS_COLUMNS = """
+    browser String,
+    os String,
+    viewport String,
+    referring_domain String,
+    utm_source String,
+    utm_medium String,
+    utm_campaign String,
+    utm_term String,
+    utm_content String,
+    country String,
+    persons_uniq_state AggregateFunction(uniq, UUID),
+    sessions_uniq_state AggregateFunction(uniq, String),
+    pageviews_count_state AggregateFunction(sum, UInt64),
+"""
+
+WEB_OVERVIEW_ORDER_BY = "(team_id, day_bucket, host, device_type)"
+WEB_STATS_ORDER_BY = "(team_id, day_bucket, host, device_type, os, browser, viewport, referring_domain, utm_source, utm_campaign, utm_medium, country)"
+
+
+def create_table_pair(base_table_name, columns, order_by, on_cluster=True):
+    """Create both a local and distributed table with the same schema"""
+    base_sql = TABLE_TEMPLATE(base_table_name, columns, order_by, on_cluster)
+    dist_sql = DISTRIBUTED_TABLE_TEMPLATE(f"{base_table_name}_distributed", base_table_name, columns)
+    return base_sql, dist_sql
+
+
+def WEB_OVERVIEW_METRICS_DAILY_SQL(table_name="web_overview_metrics_daily", on_cluster=True):
+    return TABLE_TEMPLATE(table_name, WEB_OVERVIEW_METRICS_COLUMNS, WEB_OVERVIEW_ORDER_BY, on_cluster)
+
+
+def DISTRIBUTED_WEB_OVERVIEW_METRICS_DAILY_SQL():
+    return DISTRIBUTED_TABLE_TEMPLATE(
+        "web_overview_metrics_daily_distributed", "web_overview_metrics_daily", WEB_OVERVIEW_METRICS_COLUMNS
+    )
+
+
+def WEB_STATS_DAILY_SQL(table_name="web_stats_daily", on_cluster=True):
+    return TABLE_TEMPLATE(table_name, WEB_STATS_COLUMNS, WEB_STATS_ORDER_BY, on_cluster)
+
+
+def DISTRIBUTED_WEB_STATS_DAILY_SQL():
+    return DISTRIBUTED_TABLE_TEMPLATE("web_stats_daily_distributed", "web_stats_daily", WEB_STATS_COLUMNS)
+
+
+def format_team_ids(team_ids):
+    return ", ".join(str(team_id) for team_id in team_ids)
+
+
+# This should be similar and kept in sync with what the web_overview query runner needs at posthog/hogql_queries/web_analytics/web_overview.py
+# It is ok if we have some difference in order to make the aggregations work.
+def WEB_OVERVIEW_INSERT_SQL(
```

Comment:
How easy it will be to uodate this table? Do we have a reasonable fallback? What happens if we try querying this table and it doesnt exist yet? It would be nice to have answers for all these questions (not for me, just internally in the team)

Reply:
Any error will result in the user reverting to the regular queries and fetching all events as before. 

I'm not sure how easy it will be to change the schema. We need to check what decision we'll get on the other PR, but the query to rebuild the entire history for PostHog takes about a minute, and that is what I will be working with to test the full integration.